app/src/main/java/com/github/st0rmtroop3r/weather/view_model/MainViewModel.kt

```diff
@@ -1,7 +1,46 @@
-package com.github.st0rmtroop3r.weather.ui.main
+package com.github.st0rmtroop3r.weather.view_model
 
-import android.arch.lifecycle.ViewModel
+import android.util.Log
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.github.st0rmtroop3r.weather.WeatherApp
+import com.github.st0rmtroop3r.weather.model.entities.WeatherResponse
+import com.github.st0rmtroop3r.weather.model.network.OpenWeatherMapApi
+import kotlinx.coroutines.runBlocking
+import javax.inject.Inject
 
 class MainViewModel : ViewModel() {
-    // TODO: Implement the ViewModel
+
+    private val TAG = this.javaClass.simpleName
+
+    @Inject
+    lateinit var api : OpenWeatherMapApi
+
+    private val currentWeather = MutableLiveData<WeatherResponse>()
+    private val currentWeatherError = MutableLiveData<String>()
+
+    init {
+        WeatherApp.appComponent.inject(this)
+    }
+
+    fun getCurrentWeather() = currentWeather
+
+    fun getCurrentWeatherError() = currentWeatherError
+
+    fun updateData() {
+        val deferredWeather = api.getWeather("Kyiv")
```

Comment:
Move weather receiving into usecase. Also, I'm not very familiar with coroutines, but don't you need a context here to make sure you can cancel it?

Reply:
Also the whole code block seems too complicated for plain data receiving. Maybe you should play around with suspend functions and define exceptions in usecase? I'm not sure if it will be better, but I'd suggest trying